exonum-java-binding/testkit/src/test/java/com/exonum/binding/testkit/TestKitTest.java

```diff
@@ -170,23 +171,45 @@ void createTestKitWithInvalidArtifactThrows() throws Exception {
     Class<RuntimeException> exceptionType = RuntimeException.class;
     TestKit.Builder testKitBuilder = TestKit.builder()
         .withDeployedArtifact(ARTIFACT_ID, invalidFilename)
-        .withService(ARTIFACT_ID, SERVICE_NAME, SERVICE_ID)
+        .withService(ARTIFACT_ID, SERVICE_NAME, SERVICE_ID, SERVICE_CONFIGURATION)
         .withArtifactsDirectory(artifactsDirectory);
     RuntimeException thrownException = assertThrows(exceptionType, testKitBuilder::build);
     assertThat(thrownException.getMessage())
         .contains("Failed to load the service from ", invalidFilename);
   }
 
-  // TODO: update TestService so that different configuration changes state and refactor this test
-  //  to validate that custom configuration works [ECR-3652]
   @Test
-  void createTestKitWithBuilderForSingleServiceWithCustomConfiguration() {
+  void createTestKitWithCustomConfiguration() {
+    String configurationValue = "Custom value";
+    TestConfiguration testConfiguration = TestConfiguration.newBuilder()
+        .setValue(configurationValue)
+        .build();
     try (TestKit testKit = TestKit.builder()
         .withDeployedArtifact(ARTIFACT_ID, ARTIFACT_FILENAME)
-        .withService(ARTIFACT_ID, SERVICE_NAME, SERVICE_ID, DEFAULT_CONFIGURATION)
+        .withService(ARTIFACT_ID, SERVICE_NAME, SERVICE_ID, testConfiguration)
         .withArtifactsDirectory(artifactsDirectory)
         .build()) {
-      checkTestServiceInitialization(testKit, SERVICE_NAME, SERVICE_ID);
+      TestService service = testKit.getService(SERVICE_NAME, TestService.class);
+      // Check that configuration value is used in initialization
+      Snapshot view = testKit.getSnapshot();
+      TestSchema testSchema = service.createDataSchema(view);
+      ProofMapIndexProxy<HashCode, String> testProofMap = testSchema.testMap();
+      Map<HashCode, String> testMap = toMap(testProofMap);
+      Map<HashCode, String> expected = ImmutableMap.of(
+          TestService.INITIAL_ENTRY_KEY, configurationValue);
+      assertThat(testMap).isEqualTo(expected);
+    }
+  }
+
+  @Test
+  void createTestKitWithThrowingInitialization() {
+    try (TestKit testKit = TestKit.builder()
+        .withDeployedArtifact(ARTIFACT_ID, ARTIFACT_FILENAME)
+        // Initialization without valid configuration will throw
+        .withService(ARTIFACT_ID, SERVICE_NAME, SERVICE_ID)
+        .withArtifactsDirectory(artifactsDirectory)
+        .build()) {
+      // TODO: validate the correct TestKit behaviour in case if service couldn't be started
```

Comment:
Shan't build fail with an exception?

Also, have you checked the service throws? It is protobuf, so it will happily parse messages of other types.

Reply:
Yeah, I meant throwing an exception as the "correct" behaviour.

True, I'll add special "throwing" value so that it will fail initialization.